src/Spice86.Core/Emulator/InterruptHandlers/Dos/Ems/ExpandedMemoryManager.cs

```diff
@@ -0,0 +1,1236 @@
+            EmmHandles[i] = new EmmHandle();
+        }
+
+        for (int i = 0; i < EmmMappings.Length; i++) {
+            EmmMappings[i] = new EmmMapping();
+        }
+
+        for (int i = 0; i < EmmSegmentMappings.Length; i++) {
+            EmmSegmentMappings[i] = new EmmMapping();
+        }
+
+        MemoryBlock = new MemoryBlock(MemorySizeInMb);
+
+        FillDispatchTable();
+        
+        _state.AH = EmmAllocateSystemHandle(8);
+    }
+    
+    /// <summary>
+    /// Allocates OS-dedicated handle (ems handle zero, 128kb) <br/>
+    /// This handle is never deallocated.
+    /// </summary>
+    /// <param name="logicalPage">The logical page number to allocate.</param>
+    /// <returns>Success or error code.</returns>
+    private byte EmmAllocateSystemHandle(ushort logicalPage) {
+        /* Check for enough free pages */
+        if ((GetFreeMemoryTotal() / 4) < logicalPage) {
+            if (_loggerService.IsEnabled(LogEventLevel.Warning)) {
+                _loggerService.Warning("EMS: Not enough free pages to allocate system handle. {FreePages}, {RequestedPages}", GetFreeMemoryTotal() / 4, logicalPage);
+            }
+            return EmmStatus.EmmOutOfLogicalPages;
+        }
+
+        /* Release memory if already allocated */
+        if (EmmHandles[EmmSystemHandle].LogicalPage != EmmNullHandle) {
+            ReleasePages(EmmHandles[EmmSystemHandle].MemHandle);
+        }
+        int mem = AllocatePages((ushort) (logicalPage * 4), false);
+        if (mem == 0) {
+            FailFastWithLogMessage("EMS: System handle memory allocation failure");
+        }
+        EmmHandles[EmmSystemHandle].LogicalPage = logicalPage;
+        EmmHandles[EmmSystemHandle].MemHandle = mem;
+        return EmmStatus.EmmNoError;
+    }
+
+    public uint Size => _ram.Size;
+    
+    public byte Read(uint address) {
+        uint emmPageFrameAddress = MemoryUtils.ToPhysicalAddress(EmmPageFrameSegment, 0);
+        uint relativeAddress = address - emmPageFrameAddress;
+        ThrowIfOutOfRangeForPageFrame(address, emmPageFrameAddress);
+        if (_loggerService.IsEnabled(LogEventLevel.Verbose)) {
+            _loggerService.Verbose("EMS: {@MethodName}: Read within EMM Page frame at {AbsoluteAddress} {RelativeAddress}",
+                nameof(Read), address, relativeAddress);
+        }
+        return _ram.Read(relativeAddress);
+    }
+
+    private static void ThrowIfOutOfRangeForPageFrame(uint absoluteAddress, uint emmPageFrameAddress) {
+        if (absoluteAddress < emmPageFrameAddress || absoluteAddress > emmPageFrameAddress + EmmPageFrameSize) {
+            throw new UnrecoverableException($"address {absoluteAddress} is out of range for the EMM pageframe !",
+                new ArgumentOutOfRangeException(nameof(absoluteAddress)));
+        }
+    }
+
+    public void Write(uint address, byte value) {
+        uint emmPageFrameAddress = MemoryUtils.ToPhysicalAddress(EmmPageFrameSegment, 0);
```

Comment:
Here too, I'd do:
1. translate the address to a physical page and an offset
2. look up which logical page number is currently mapped to that physical page number
3. access the Ram belonging to that logical page number at the offset calculated at 1.

Something like
```pseudocode
var physicalPageNr = (address - emmPageFrameAddress) % EmmPageSize ;
var offset = (address - emmPageFrameAddress) - (physicalPageNr  * EmmPageSize);
var logicalPageNr = pageMappings[handle][physicalPageNr];
LogicalPageRams[logicalPageNr].Write(offset, value);
```

Reply:
I need to rework the EmmMapping structure, but I'm getting closer to what you suggested.